Submission client major refactors

mccheah opened the pull request:
Looking to split this out into multiple chunks if possible, but this is a brainstorming project on how the submission process could be cleaned up.

This commit contains all of the following refactor ideas:

- Break apart the Client into several components: (1) Configuring the Kubernetes client, (2) Interacting with the API server to create and clean up Kubernetes resources, and (3) submitting the application to the driver submission server. `KubernetesDriverLifecycle` is a composition of the latter two responsibilities, but only implements (2) and leaves the implementation details of (3) to `KubernetesDriverSubmitter`. Client is now only a proxy to `KubernetesDriverLifecycle` with the thin wrapping of watching the driver pod's status, constructing all of the top-level components, and validating the existence of provided local files.
- Constructing Kubernetes components is now done through creating models via local builders, and then shipping these built models through to the Kubernetes Java client.
- Component readiness now uses Fabric8's `ReadinessWatcher` implementation that can be used across any kind of resource, as opposed to the previous approach that provided a specific `Watcher` type for every resource. Waiting for readiness has been pulled out to a separate class. The new implementation is inspired by Fabric8's but has more robust handling of a race condition that would cause Fabric8's implementation to hang indefinitely. This mode of operation is made easier by using local builders as mentioned earlier.
- `KubernetesResourceCleaner` is no longer used since we now bulk-create resources, and thus having an extra class is superfluous. There is now a utility that allows wrapping a piece of logic into both a shutdown hook and a catch block, where if the client is shut down prematurely the cleanup block should still run. We use this to clean up Kubernetes resources; if the application submission is halted either by a JVM shutdown or an exception, the resources should still be deleted from the cluster.

mccheah commented:
@foxish @robert3005 @iyanuobidele @ssuchter @ash211 @kimoonkim

The diff is a little large, but I hope that we can discuss the four refactor ideas that I have presented here. I will likely repost this work as several smaller diffs, most likely in the four categories mentioned here. But how does this approach sound from a bigger picture perspective?

ash211 commented:
Please rebase onto `branch-2.1-kubernetes` and send this PR into that branch instead of `k8s-support-alternate-incremental` which is now deprecated.

mccheah committed:
Submission client major refactors.

This commit contains all of the following refactor ideas:

- Break apart the Client into several components: (1) Configuring the
Kubernetes client, (2) Interacting with the API server to create and clean
up Kubernetes resources, and (3) submitting the application to the driver
submission server. KubernetesDriverLifecycle is a composition of the
latter two responsibilities, but only implements (2) and leaves the
implementation details of (3) to KubernetesDriverSubmitter. Client is now
only a proxy to KubernetesDriverLifecycle with the thin wrapping of watching
the driver pod's status, constructing all of the top-level components, and
validating the existence of provided local files.
- Constructing Kubernetes components is now done through creating models
via local builders, and then shipping these built models through to the
Kubernetes Java client.
- Component readiness now uses Fabric8's ReadinessWatcher implementation
that can be used across any kind of resource, as opposed to the previous
approach that provided a specific Watcher type for every resource.
Waiting for readiness has been pulled out to a separate class. The new
implementation is inspired by Fabric8's but has more robust handling of
a race condition that would cause Fabric8's implementation to hang
indefinitely. This mode of operation is made easier by using local
builders as mentioned earlier.
- KubernetesResourceCleaner is no longer used since we now bulk-create
resources, and thus having an extra class is superfluous.
There is now a utility that allows wrapping a piece of logic into both a
shutdown hook and a catch block, where if the client is shut down prematurely
the cleanup block should still run. We use this to clean up Kubernetes
resources; if the application submission is halted either by a JVM
shutdown or an exception, the resources should still be deleted from the
cluster.

mccheah committed:
Merge remote-tracking branch 'apache-spark-on-k8s/branch-2.1-kubernetes' into refactor-kubernetes-resource-management

mccheah committed:
Fix compilation error

mccheah commented:
I'm going to abandon this change since submission V2 seems more promising rather than trying to clean up something we probably won't use moving forward.
